From e31e1609f56e190e71a7cc152df3dfdf8df0636c Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Tue, 17 Feb 2026 10:45:07 -0500
Subject: [PATCH] Require __NR_getrandom to be either available or polyfilled

Dating to when getrandom was very new, we wouldn't use getrandom if it
wasn't defined in either headers or in the fillin header. Worse, it was
defined in headers but not the fillin header, we still wouldn't use it!
(Though we had values for all supported architectures, so this is
largely moot.)

However, after
https://boringssl-review.googlesource.com/c/boringssl/+/86487, builds
without USE_NR_getrandom broken anyway. (The namespace import was in the
wrong file.) Given those were broken anyway, just remove that polyfill.

I haven't quite gone as far as to require __NR_getrandom be available in
headers, but I suspect we can at this point. (We have a bunch of stale
glibc headers around, but IIRC those come from Linux headers rather than
glibc headers.) Left a TODO to run a TGP on that.

Bug: 446280907
Change-Id: I3b4d4422050e0b5d47f8909d6ad759fcaa144a86
Reviewed-on: https://boringssl-review.googlesource.com/c/boringssl/+/89508
Commit-Queue: Lily Chen <chlily@google.com>
Reviewed-by: Lily Chen <chlily@google.com>
Auto-Submit: David Benjamin <davidben@google.com>
---
 .../src/crypto/rand/getrandom_fillin.h        | 20 ++++++++-----------
 .../boringssl/src/crypto/rand/urandom.cc      | 11 ----------
 .../boringssl/src/crypto/rand/urandom_test.cc |  8 +++-----
 3 files changed, 11 insertions(+), 28 deletions(-)

diff --git a/third_party/boringssl/src/crypto/rand/getrandom_fillin.h b/third_party/boringssl/src/crypto/rand/getrandom_fillin.h
index 0057931b5f..186fff18c5 100644
--- a/third_party/boringssl/src/crypto/rand/getrandom_fillin.h
+++ b/third_party/boringssl/src/crypto/rand/getrandom_fillin.h
@@ -22,6 +22,7 @@
 
 #include <sys/syscall.h>
 
+// TODO(crbug.com/446280907): Can we assume that __NR_getrandom is defined?
 #if defined(OPENSSL_X86_64)
 #define EXPECTED_NR_getrandom 318
 #elif defined(OPENSSL_X86)
@@ -36,23 +37,18 @@
 #define EXPECTED_NR_getrandom 278
 #endif
 
-#if defined(EXPECTED_NR_getrandom)
-#define USE_NR_getrandom
+#if !defined(__NR_getrandom) && defined(EXPECTED_NR_getrandom)
+#define __NR_getrandom EXPECTED_NR_getrandom
+#endif
 
-#if defined(__NR_getrandom)
+#if !defined(__NR_getrandom)
+#error "__NR_getrandom was not defined"
+#endif
 
-#if __NR_getrandom != EXPECTED_NR_getrandom
+#if defined(EXPECTED_NR_getrandom) && __NR_getrandom != EXPECTED_NR_getrandom
 #error "system call number for getrandom is not the expected value"
 #endif
 
-#else  // __NR_getrandom
-
-#define __NR_getrandom EXPECTED_NR_getrandom
-
-#endif  // __NR_getrandom
-
-#endif  // EXPECTED_NR_getrandom
-
 #if !defined(GRND_NONBLOCK)
 #define GRND_NONBLOCK 1
 #endif
diff --git a/third_party/boringssl/src/crypto/rand/urandom.cc b/third_party/boringssl/src/crypto/rand/urandom.cc
index 0282926306..afcd462626 100644
--- a/third_party/boringssl/src/crypto/rand/urandom.cc
+++ b/third_party/boringssl/src/crypto/rand/urandom.cc
@@ -33,8 +33,6 @@
 #include "getrandom_fillin.h"
 
 
-#if defined(USE_NR_getrandom)
-
 using namespace bssl;
 
 #if defined(OPENSSL_MSAN)
@@ -60,8 +58,6 @@ static ssize_t boringssl_getrandom(void *buf, size_t buf_len, unsigned flags) {
   return ret;
 }
 
-#endif  // USE_NR_getrandom
-
 // kHaveGetrandom in |urandom_fd| signals that |getrandom| or |getentropy| is
 // available and should be used instead.
 static const int kHaveGetrandom = -3;
@@ -75,7 +71,6 @@ static CRYPTO_once_t rand_once = CRYPTO_ONCE_INIT;
 // requested. This is the only function that modifies |urandom_fd|, which may be
 // read safely after calling the once.
 static void init_once() {
-#if defined(USE_NR_getrandom)
   int have_getrandom;
   uint8_t dummy;
   ssize_t getrandom_ret =
@@ -98,7 +93,6 @@ static void init_once() {
     urandom_fd = kHaveGetrandom;
     return;
   }
-#endif  // USE_NR_getrandom
 
   // FIPS builds must support getrandom.
 #if defined(BORINGSSL_FIPS)
@@ -136,12 +130,7 @@ void bssl::CRYPTO_sysrand(uint8_t *out, size_t len) {
     ssize_t r;
 
     if (urandom_fd == kHaveGetrandom) {
-#if defined(USE_NR_getrandom)
       r = boringssl_getrandom(out, len, 0);
-#else  // USE_NR_getrandom
-      fprintf(stderr, "urandom fd corrupt.\n");
-      abort();
-#endif
     } else {
       do {
         r = read(urandom_fd, out, len);
diff --git a/third_party/boringssl/src/crypto/rand/urandom_test.cc b/third_party/boringssl/src/crypto/rand/urandom_test.cc
index 1b0ad25a08..21cd329cf4 100644
--- a/third_party/boringssl/src/crypto/rand/urandom_test.cc
+++ b/third_party/boringssl/src/crypto/rand/urandom_test.cc
@@ -28,9 +28,8 @@
 #include "internal.h"
 
 
-#if (defined(OPENSSL_X86_64) || defined(OPENSSL_AARCH64)) &&               \
-    !defined(BORINGSSL_SHARED_LIBRARY) && defined(OPENSSL_RAND_URANDOM) && \
-    defined(USE_NR_getrandom)
+#if (defined(OPENSSL_X86_64) || defined(OPENSSL_AARCH64)) && \
+    !defined(BORINGSSL_SHARED_LIBRARY) && defined(OPENSSL_RAND_URANDOM)
 
 #include <elf.h>
 #include <linux/random.h>
@@ -790,5 +789,4 @@ int main(int argc, char **argv) {
   return 0;
 }
 
-#endif  // (X86_64 || AARCH64) && !SHARED_LIBRARY &&
-        // !UNSAFE_DETERMINISTIC_MODE && USE_NR_getrandom
+#endif  // (X86_64 || AARCH64) && !SHARED_LIBRARY && !UNSAFE_DETERMINISTIC_MODE
-- 
2.51.0

